pkg/gcpbuildpack/filepath.go

```diff
@@ -45,6 +54,10 @@ func (ctx *Context) HasAtLeastOne(pattern string) (bool, error) {
 	}
 
 	if err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if info.IsDir() && DETECTOR_IGNORED_DIRS[info.Name()] {
```

Comment:
I think we wouldn't want something like this in a general purpose HasAtLeastOne(...) function as this will affect all callers, is surprising behavior to developers, and will be difficult to walk back in the future when it proves to be too restrictive for some use cases. Effectively, it's too hard-coded, and too specific to certain languages for this general purpose package. For example, even with only two directories ignored, I'm not sure that ignoring 'vendor' across all languages is the right call.

As an alternative, I can see adding another version of HasAtLeastOne(...) that allows callers to pass in a filter function or potentially to add optional filter argument to this function. The PHP buildpack / package can be updated to use a HasAtLeastOne(...) that filters out node_modules.

Reply:
This would be a way to do an optional filter (a variable length list of functions): 

  https://github.com/GoogleCloudPlatform/buildpacks/blob/main/pkg/gcpbuildpack/gcpbuildpack.go#L144

So it would be something like this:

```
type InclusionFilter func(string) bool

func (ctx *Context) HasAtLeastOne(pattern string, filters ...InclusionFilter) (bool, error) {
   for _, f := range filters {
     // check to see if the given path should be included by invoking f(path)
   }
}
```

Another way to do this would be by having this method call another method that takes an optional function parameter, note this avoids code duplication:

```

type InclusionFilter func(string) bool

func (ctx *Context) HasAtLeastOne(pattern string) (bool, error) {
  return ctx.HasAtLeastOneFiltered(pattern, nil)
}

// HasAtLeastOneFiltered is a pass through for filepath.Glob(...) it returns true if there is at least one 
// file which matches the search pattern and is included by `filter`
func (ctx *Context) HasAtLeastOneFiltered(pattern string, filter InclusionFilter) (bool, error)
{
   // insert all the code from the existing HasAtLeastOne here and use the `filter` if it is NOT `nil`
}
```